tests/MongoDB.Driver.Core.Tests/Core/Connections/ConnectionInitializerTests.cs

```diff
@@ -67,6 +69,63 @@ public void InitializeConnection_should_acquire_connectionId_from_isMaster_respo
             result.ConnectionId.ServerValue.Should().Be(1);
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void InitializeConnection_should_speculatively_authenticate_with_SCRAM_SHA(
+            [Values("default", "SCRAM-SHA-256", "SCRAM-SHA-1")] string authenticatorType,
+            [Values(false, true)] bool async)
+        {
+            var isMasterReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, connectionId : 1 }"));
+            var buildInfoReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, version : \"4.2.0\" }"));
+            var credentials = new UsernamePasswordCredential(
+                source: "Voyager", username: "Seven of Nine", password: "Omega-Phi-9-3");
+            var authenticator = CreateAuthenticator();
+            var connectionSettings = new ConnectionSettings(new[] { authenticator });
+            var connection = new MockConnection(__serverId, connectionSettings, eventSubscriber: null);
+            connection.EnqueueReplyMessage(isMasterReply);
+            connection.EnqueueReplyMessage(buildInfoReply);
+
+            // We expect authentication to fail since we have not enqueued the expected authentication replies
+            try
+            {
+                if (async)
+                {
+                    _subject.InitializeConnectionAsync(connection, CancellationToken.None).GetAwaiter().GetResult();
+                }
+                else
+                {
+                     _subject.InitializeConnection(connection, CancellationToken.None);
+                }
+            }
+            catch (InvalidOperationException ex)
+            {
+                ex.Message.Should().Be("Queue empty.");
+            }
+
+            var sentMessages = connection.GetSentMessages();
+            var isMasterQuery = (QueryMessage)sentMessages[0];
+            var isMasterDocument = isMasterQuery.Query;
+            isMasterDocument.Should().Contain("speculativeAuthenticate");
+            var speculativeAuthenticateDocument = isMasterDocument["speculativeAuthenticate"].AsBsonDocument;
+            speculativeAuthenticateDocument.Should().Contain("mechanism");
+            var expectedMechanism = new BsonString(
+                authenticatorType == "default" ? "SCRAM-SHA-256" : authenticatorType);
+            speculativeAuthenticateDocument["mechanism"].Should().Be(expectedMechanism);
+
+            IAuthenticator CreateAuthenticator()
```

Comment:
Can't make this static local since we're not on C#8. https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#static-local-functions